src/System.Private.CoreLib/shared/System/Runtime/CompilerServices/AsyncValueTaskMethodBuilderT.cs

```diff
@@ -0,0 +1,472 @@
+        private abstract class StateMachineBox :
+            IValueTaskSource<TResult>, IValueTaskSource
+        {
+            /// <summary>Maximum number of boxes that are allowed to be cached.</summary>
+            protected static readonly int s_maxCacheSize = Environment.ProcessorCount * 4; // arbitrary for now
+
+            /// <summary>A delegate to the MoveNext method.</summary>
+            protected Action? _moveNextAction;
+            /// <summary>Captured ExecutionContext with which to invoke MoveNext.</summary>
+            public ExecutionContext? Context;
+            /// <summary>Implementation for IValueTaskSource interfaces.</summary>
+            protected ManualResetValueTaskSourceCore<TResult> _valueTaskSource;
+
+            /// <summary>Completes the box with a result.</summary>
+            /// <param name="result">The result.</param>
+            public void SetResult(TResult result) =>
+                _valueTaskSource.SetResult(result);
+
+            /// <summary>Completes the box with an error.</summary>
+            /// <param name="error">The exception.</param>
+            public void SetException(Exception error) =>
+                _valueTaskSource.SetException(error);
+
+            /// <summary>Gets the status of the box.</summary>
+            public ValueTaskSourceStatus GetStatus(short token) => _valueTaskSource.GetStatus(token);
+
+            /// <summary>Schedules the continuation action for this box.</summary>
+            public void OnCompleted(Action<object?> continuation, object? state, short token, ValueTaskSourceOnCompletedFlags flags) =>
+                _valueTaskSource.OnCompleted(continuation, state, token, flags);
+
+            /// <summary>Gets the result of the operation.</summary>
+            public abstract TResult GetResult(short token);
+
+            /// <summary>Gets the result of the operation.</summary>
+            void IValueTaskSource.GetResult(short token) => GetResult(token);
+
+            /// <summary>Gets the current version number of the box.</summary>
+            public short Version => _valueTaskSource.Version;
+        }
+
+        /// <summary>Provides a strongly-typed box object based on the specific state machine type in use.</summary>
+        private sealed class StateMachineBox<TStateMachine> :
+            StateMachineBox, IAsyncStateMachineBox, IThreadPoolWorkItem
+            where TStateMachine : IAsyncStateMachine
+        {
+            /// <summary>Delegate used to invoke on an ExecutionContext when passed an instance of this box type.</summary>
+            private static readonly ContextCallback s_callback = ExecutionContextCallback;
+            /// <summary>Lock used to protected the shared cache of boxes.</summary>
+            private static SpinLock s_cacheLock = new SpinLock(enableThreadOwnerTracking: false); // mutable struct; do not make this readonly
```

Comment:
Any reason this isn't just a lock with a Monitor.TryEnter?

Reply:
It's going to just be an interlocked on an int; there's no reason to use either a monitor or a spinlock given where I ended up with it. Just haven't revised it yet.

Reply:
Ah ok! Makes sense.

Reply:
Changed.  It now just uses Interlocked.CompareExchange.  Made a few other streamlining tweaks as well.